CI: split PR workflows, add paths

markgoddard opened the pull request:
- CI: Split .github/workflows/stackhpc-pull-request.yml into tox and aio workflows
- CI: Add paths to stackhpc-pull-request-tox-aio.yml workflow

This should reduce our CI overheads by only running aios when necessary. It will also make it easier to merge some changes.

markgoddard commented:
Cancelled CI jobs because we can see they were triggered correctly.

Alex-Welsh commented on .github/workflows/stackhpc-pull-request-aio.yml:
```diff
@@ -1,39 +1,48 @@
+      - '!etc/kayobe/bmc.yml'
+      - '!etc/kayobe/cephadm.yml'
+      - '!etc/kayobe/containers/**'
+      - '!etc/kayobe/environments/aufn-ceph/**'
+      - '!etc/kayobe/environments/ci-builder/**'
+      - '!etc/kayobe/environments/ci-multinode/**'
```
```suggestion
      - '!etc/kayobe/environments/**'
      - 'etc/kayobe/environments/ci-aio/**'
```

markgoddard replied:
Like it

markgoddard commented:
I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(

Oh GH actions, you are no zuul.

Alex-Welsh commented:
> I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(
> 
> Oh GH actions, you are no zuul.

Can we fake running it and immediately return a success?

markgoddard commented:
> > I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(
> > Oh GH actions, you are no zuul.
> 
> Can we fake running it and immediately return a success?

Possibly, but it would be a bit misleading and would need something other than the `paths` selector

markgoddard committed:
CI: Skip aio jobs based on changed files

We don't always want to run all aio jobs - it can be wasteful of CI
resources and slows down merging changes.

This change uses the dorny/paths-filter action to detect changed files
and skip jobs as appropriate.

Note that we can't use the workflow-level paths attribute since this
would skip the workflow entirely, and would prevent us from making the
aio jobs required to pass (a skip counts as a pass).

markgoddard committed:
CI: Remove unnecessary conditionals

The job dependency should handle this, with the condition applied to the first job.

markgoddard committed:
DNM: test path filter

Alex-Welsh commented:
How did the testing for this go? does it need more work?

markgoddard commented:
> How did the testing for this go? does it need more work?

Looks like it doesn't accept the skipped jobs :(

jovial commented:
> > How did the testing for this go? does it need more work?
> 
> Looks like it doesn't accept the skipped jobs :(

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks. Seems like the only option is skip all jobs within a workflow rather than the workflow itself :(

markgoddard committed:
Try adding condition to all jobs

markgoddard commented:
> > > How did the testing for this go? does it need more work?
> > 
> > 
> > Looks like it doesn't accept the skipped jobs :(
> 
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks. Seems like the only option is skip all jobs within a workflow rather than the workflow itself :(

That's what it's doing already. I was relying on job dependencies though and have just tried adding the condition to every job. Could you approve again?

It does look like it's just not recognising the jobs though as it lists them as skipped and expected.

markgoddard commented:
I managed to tweak the branch protection rules to require the calling workflow's job rather than the called workflow, and this seems to now recognise the skipped results as success. I guess it never gets to the called workflow if the calling job is skipped.

Just need to confirm that this still requires the jobs to pass when they run :crossed_fingers:

markgoddard commented:
> I managed to tweak the branch protection rules to require the calling workflow's job rather than the called workflow, and this seems to now recognise the skipped results as success. I guess it never gets to the called workflow if the calling job is skipped.
> 
> Just need to confirm that this still requires the jobs to pass when they run 🤞

Lame, it doesn't work:

![image](https://github.com/stackhpc/stackhpc-kayobe-config/assets/5795873/2578908a-8b8a-4864-b1d8-927f48b8db47)

I'll revert the required status checks change.

markgoddard committed:
Try moving condition to called workflows

markgoddard commented:
I tried pushing the condition into the called workflow, and this worked for the Build Kayobe image check, but not for the aio checks because they depend on the image job and so don't run.

markgoddard committed:
Try always

markgoddard committed:
DNM: test broken branch

markgoddard committed:
DNM: now fix it

markgoddard committed:
Add dep

markgoddard committed:
Revert "DNM: now fix it"

This reverts commit bd64f5af80f66c651620132655e57b21aa05b7fc.

markgoddard committed:
Revert "Try always"

This reverts commit b075159e47f2b257103569fe6e43d72d4d4cd0a6.

markgoddard commented:
Started a [discussion](https://github.com/orgs/community/discussions/72708) on this.

markgoddard commented:
Broken build kayobe image job
No always()
Status not recognised, but not a problem since the build failed.

markgoddard committed:
Revert "Revert "DNM: now fix it""

This reverts commit a3977beabca8b1546b7e91882ab3c2fe3891d6c9.

markgoddard commented:
Changed file triggers jobs
Fixed build kayobe image job
No always()
Status recognised.

markgoddard committed:
DNM: Remove changes

markgoddard commented:
No changed files so jobs not triggered
No always()
Status not recognised.

markgoddard committed:
Revert "Revert "Try always""

This reverts commit 088677a317545cba4c82ad8f587fcf07a770ae63.

markgoddard commented:
No changed files so jobs not triggered
Using always()
Status recognised.

markgoddard committed:
Test changes

markgoddard commented:
Changed file triggers jobs
Working build kayobe image job
Using always()
Status recognised.

markgoddard committed:
DNM: Break kayobe image build

markgoddard commented:
Changed file triggers jobs
Failing build kayobe image job
Using always()
Status recognised.
aio jobs run but should be skipped

markgoddard committed:
DNM: Debug job result

markgoddard committed:
unbreak kayobe image

markgoddard committed:
DNM

markgoddard commented:
No changed files so jobs not triggered
Using ${{ ! failure() }}
Status recognised.

Using always() && needs.build-kayobe-image.result == 'success'
Status not recognised: result == 'skipped'.

markgoddard commented:
No changed files so jobs not triggered
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
Status recognised.

markgoddard committed:
add changes

markgoddard commented:
Changed file triggers jobs
Failing build kayobe image job
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
aio jobs skipped & not recognised as success

markgoddard committed:
fix kayobe build

markgoddard commented:
Changed file triggers jobs
Working build kayobe image job
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
aio jobs run & recognised as success

markgoddard commented:
Looks like either of these should do the trick

markgoddard committed:
if failure

markgoddard committed:
remove reusable if

markgoddard committed:
Remove changed file

markgoddard commented:
This PR is a bit of a mess. Closing in favour of https://github.com/stackhpc/stackhpc-kayobe-config/pull/726
